Add NumberView and HashView components

helenamartins81 opened the pull request:
**Why:** 

- Currently, it is not possible to convert numbers into different units (wei, gwei, eth) or different formats (decimal to hex)  (Issue #695) 

**How:**
- Added the NumberView and HashView components to enable conversion between units in numbers, copy to clipboard functionality, and format switching between decimal and hex.  
gui/src/components/NumberView.tsx
gui/src/components/HashView.tsx

**Preview**:
![image](https://github.com/ethui/ethui/assets/14261331/46dd8bd6-25d5-49c0-a7b8-5ee2cc5d22c0)

naps62 commented on gui/src/routes/_home.home/transactions.lazy.tsx:
```diff
@@ -223,24 +219,40 @@ function Details({ tx, chainId }: DetailsProps) {
-            value={`${fullTx.maxFeePerGas && formatGwei(BigInt(fullTx.maxFeePerGas))} gwei`}
+            value={
+              fullTx.maxFeePerGas ? (
+                <NumberView value={Number(fullTx.maxFeePerGas)} />
```
this seems incorrect. `maxFeePerGas` and `maxPriorityFeePerGas` are already expressed in `gwei`, but that information is lost here

I'd expect something along the lines of

`<NumberView value={fullTx.maxFeePerGas} unit="gwei" />`

where the component would be smart enough to show me the conversions from there

naps62 commented on gui/src/components/NumberView.tsx:
```diff
@@ -0,0 +1,85 @@
+      setDisplayValue(value.toString());
+    }
+    setMenuAnchor(null);
+  };
```
a "Change Format" button on the UI is very unclear as to what its purpose is.
from seeing the code, seems to toggle between hex and decimal notation. but for a user that won't be clear at all

naps62 commented:
Overall a nice effort, but I feel this somewhat misses the point.

we can't blindly wrap values in a `NumberView` and offer a couple of static conversions. Some values already come in wei, others come in gwei or eth. We need to be smart about the parameters, and how to convert between them (viem already has functions to deal with all this, as long as we know the decimals)

What I was expecting, was not a context menu, but a modal showing the different values the number can take, so I can copy to clipboard the one I'm interested in

helenamartins81 commented:
**Update:**

Removed the menu and added a modal with all the conversions (hashes only display hex and decimal values).
In NumbersView, adjusted the conversion process to consider the initial value unit.

**Preview:**

![image](https://github.com/ethui/ethui/assets/14261331/a4766fb3-a6f3-4f96-865b-31dfc61d2714)

naps62 commented on gui/src/components/NumberView.tsx:
```diff
@@ -0,0 +1,123 @@
+    eth: "",
+    decimal: "",
+    hex: "",
+  });
```
wei/gwei/eth are not in the same group as decimal/hex. these are different axis
e.g.: I may want to view the same gwei value in decimal or hex

also, there are many more units other than these 3: https://www.gemini.com/cryptopedia/satoshi-value-gwei-to-ether-to-wei-converter-eth-gwei#section-ethereum-denominations-ether-to-wei-gwei-to-ether-more
for each of these, you can just map it to a number of decimal places, and use `formatUnits` from viem instead of manually handling each of them

helenamartins81 commented:
**Update**:

In NumbersView, moved the hexadecimal format to another popover. Now, it is possible to see and copy the conversion to hexadecimal for each unit. I also added more units and adjusted the conversion process using formatUnits from viem.

**Preview:**

![image](https://github.com/user-attachments/assets/de420df3-5ee5-469c-8ca6-4736693f0519)
